skaffold-plugins-core/src/main/java/com/google/cloud/tools/skaffold/command/Skaffold.java

```diff
@@ -0,0 +1,219 @@
+import com.google.common.io.ByteStreams;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/** Runs {@code skaffold} commands. */
+public class Skaffold {
+
+  /** The location to store {@code skaffold} if auto-downloading it. */
+  private static final Path CACHED_SKAFFOLD_LOCATION =
+      UserCacheHome.getCacheHome().resolve("skaffold");
+
+  /** The location to store the digest for {@code skaffold} if auto-downloading it. */
+  private static final Path CACHED_SKAFFOLD_DIGEST_LOCATION =
+      CACHED_SKAFFOLD_LOCATION.resolveSibling("skaffold.sha256");
+
+  /**
+   * Initializes {@link Skaffold} with a custom path to the {@code skaffold} executable.
+   *
+   * @param executablePath the path to {@code skaffold}
+   * @return a new {@link Skaffold}
+   */
+  public static Skaffold atPath(Path executablePath) {
+    return new Skaffold(executablePath.toString());
+  }
+
+  /**
+   * Initializes {@link Skaffold} with a managed {@code skaffold} executable.
+   *
+   * @return a new {@link Skaffold}
+   * @throws IOException if an I/O exception occurred
+   */
+  public static Skaffold init() throws IOException {
+    ensureSkaffoldIsLatestVersion(CACHED_SKAFFOLD_LOCATION, CACHED_SKAFFOLD_DIGEST_LOCATION);
+    return new Skaffold(CACHED_SKAFFOLD_LOCATION.toString());
+  }
+
+  @VisibleForTesting
+  static void ensureSkaffoldIsLatestVersion(
+      Path cachedSkaffoldLocation, Path cachedSkaffoldDigestLocation) throws IOException {
+    if (Files.exists(cachedSkaffoldLocation)) {
+      // Checks if the digest is up-to-date and redownloads skaffold if not.
+      Path temporaryDigestFile = Files.createTempFile("", "");
+      temporaryDigestFile.toFile().deleteOnExit();
+      SkaffoldDownloader.downloadLatestDigest(temporaryDigestFile);
+      byte[] latestDigest = Files.readAllBytes(temporaryDigestFile);
+      if (Files.exists(cachedSkaffoldDigestLocation)) {
+        byte[] storedDigest = Files.readAllBytes(cachedSkaffoldDigestLocation);
+        if (Arrays.equals(storedDigest, latestDigest)) {
+          return;
+        }
+      }
+      Files.write(cachedSkaffoldDigestLocation, latestDigest);
+    }
+
+    SkaffoldDownloader.downloadLatest(cachedSkaffoldLocation);
+  }
+
+  private static Callable<Void> redirect(InputStream inputStream, OutputStream outputStream) {
+    return () -> {
+      try (InputStream inputStream1 = inputStream) {
```

Comment:
This implicitly closes the provided inputStream?  Closing streams is usually the responsibility of the code that opened the stream.